File: src/scripts/shared/git-utils.mjs
/**
 * Git Utilities
 * 
 * Common git-related functions for build scripts.
 */

import { access } from 'fs/promises';
import { join, dirname } from 'path';

/**
 * Find git root by looking for .git directory
 * 
 * @param {string} startPath - Path to start searching from
 * @returns {Promise<string>} Absolute path to git root
 * @throws {Error} If git root not found
 */
export async function findGitRoot(startPath) {
  let currentPath = startPath;
  
  while (true) {
    try {
      await access(join(currentPath, '.git'));
      return currentPath;
    } catch {
      const parentPath = dirname(currentPath);
      if (parentPath === currentPath) {
        throw new Error('Git root not found - not in a git repository?');
      }
      currentPath = parentPath;
    }
  }
}

File: src/scripts/shared/GitUtils.cs
using System.Xml.Linq;

namespace BuildScripts.Shared;

/// <summary>
/// Git-related utility functions
/// </summary>
public static class GitUtils
{
    /// <summary>
    /// Find git repository root by looking for .git directory
    /// </summary>
    /// <param name="startPath">Path to start searching from</param>
    /// <returns>Absolute path to git root</returns>
    /// <exception cref="DirectoryNotFoundException">If git root not found</exception>
    public static string FindGitRoot(string? startPath = null)
    {
        var currentPath = startPath ?? Directory.GetCurrentDirectory();
        
        while (true)
        {
            if (Directory.Exists(Path.Combine(currentPath, ".git")))
            {
                return currentPath;
            }
            
            var parentPath = Directory.GetParent(currentPath)?.FullName;
            if (parentPath == null || parentPath == currentPath)
            {
                throw new DirectoryNotFoundException("Git root not found - not in a git repository?");
            }
            
            currentPath = parentPath;
        }
    }
}
